File: chip8/keyboard.go
package chip8

const (
	// NumKeys is the number of keys in the CHIP-8 hexadecimal keyboard.
	NumKeys = 0x10
)

// Keyboard is the CHIP-8 hexadecimal keyboard.
type Keyboard struct {
	keys    [NumKeys]bool
	polling bool
	lastKey int
}

func NewKeyboard() *Keyboard {
	return &Keyboard{
		keys:    [NumKeys]bool{},
		polling: false,
		lastKey: -1,
	}
}

func (k *Keyboard) IsPressed(key uint8) bool {
	return k.keys[key]
}

func (k *Keyboard) Set(key uint8, pressed bool) {
	if k.polling && k.lastKey == -1 {
		k.lastKey = int(key)
	}
	k.keys[key] = pressed
}

func (k *Keyboard) Poll() (uint8, bool) {
	if k.polling && k.lastKey >= 0 {
		key := uint8(k.lastKey)
		k.polling = false
		k.lastKey = -1
		return key, true
	} else {
		k.polling = true
		return 0, false
	}
}

File: src/chip8/keyboard.rs
pub struct Keyboard {
    keys: [bool; 0x10],
    polling: bool,
    last_pressed: i16,
}

impl Keyboard {
    pub fn new() -> Self {
        Self {
            keys: [false; 0x10],
            polling: false,
            last_pressed: -1,
        }
    }

    pub fn set(&mut self, key: u8, value: bool) {
        if value && self.last_pressed < 0 {
            self.last_pressed = key as i16
        }
        self.keys[key as usize] = value;
    }

    pub fn is_pressed(&self, key: u8) -> bool {
        self.keys[key as usize]
    }

    pub fn poll(&mut self) -> (u8, bool) {
        if self.polling && self.last_pressed >= 0 {
            let key = self.last_pressed as u8;
            self.polling = false;
            self.last_pressed = -1;
            (key, true)
        } else {
            self.polling = true;
            (0, false)
        }
    }
}
